core/src/main/scala/cromwell/util/JsonEditor.scala

```diff
@@ -72,21 +73,77 @@ object JsonEditor {
 
   def logs(json: Json): Json = includeJson(json, NonEmptyList.of("stdout", "stderr", "backendLogs", "id"))
 
+  implicit class EnhancedJson(val json: Json) extends AnyVal {
+    def rootWorkflowId: Option[WorkflowId] = {
+      for {
+        o <- json.asObject
+        id <- o.kleisli("id")
+        s <- id.asString
+      } yield WorkflowId.fromString(s)
+    }
+  }
+
   /**
-    * We only return labels associated with top-level workflows.  Subworkflows don't include labels (as of 7/26/19).
-    *
-    * Thus this method puts the labels as a top-level field.
+    * In-memory upsert of labels into the base Json, handling root and sub workflows appropriately.
     *
     * @param json json blob with or without "labels" field
-    * @param labels a map of labels one would like to apply to a workflow json
+    * @param databaseLabels a map of workflow IDs to maps of labels one would like to apply to a workflow json
     * @return json with labels merged in.  Any prior non-object "labels" field will be overwritten and any object fields will be merged together and - again - any existing values overwritten.
     */
-  def augmentLabels(json: Json, labels: Map[String, String]): Json = {
-    val newData: Json = Json.fromFields(labels.safeMapValues(Json.fromString))
-    val newObj: Json = Json.fromFields(List(("labels", newData)))
-    //in the event of a key clash, the values in "newObj" will be favored over "json"
-    json deepMerge newObj
-  }
+  def updateLabels(json: Json, databaseLabels: Map[WorkflowId, Map[String, String]]): Json = {
+    val subWorkflowMetadataKey = "subWorkflowMetadata"
+
+    def doUpdateWorkflow(workflowJson: Json): Json = {
+      val id: String = workflowJson.rootWorkflowId.
+        getOrElse(throw new RuntimeException(s"did not find workflow id in ${workflowJson.printWith(Printer.spaces2)}")).toString
+
+      // Look for an optional JsonObject keyed by "calls".
+      val callsObject: Option[JsonObject] = for {
+        wo <- workflowJson.asObject
+        callsJson <- wo("calls")
+        co <- callsJson.asObject
+      } yield co
+
+      val workflowWithUpdatedCalls: Json = callsObject match {
+        // If there were no calls just return the workflow JSON unmodified.
+        case None => workflowJson
+        case Some(calls) =>
+          val updatedCallsObject = calls.mapValues {
+            // The Json (a JSON array, really) corresponding to the array of call objects for a call name.
+            callValue: Json =>
+              // The object above converted to a List[Json].
+              val callArray: List[Json] = callValue.asArray.toList flatMap { _.toList }
```

Comment:
`callValue.asArray.toVector.flatten` feels more readable, and lets you keep it as a `Vector` rather than re-casting to a `List`